File: C/S15_GRAPH/L1_BASIC/4_bfs.c
#include<stdio.h>
#include<stdlib.h>
struct node{
    int data;
    struct node*next;
};
struct node*createnode(int data){
    struct node*newnode=(struct node*)malloc(sizeof(struct node));
    newnode->data=data;
    newnode->next=NULL;
    return newnode;
}
struct queue{
    int rear;
    int front;
    int *arr;
};
struct queue *createque(int size){
    struct queue *q=(struct queue*)malloc(sizeof(struct queue));
    q->arr=(int*)malloc(sizeof(int)*size);
    q->rear=-1;
    q->front=0;
    return q;
}
struct graph {
    int numsize;
    struct node **adjlist;
    int *visited;

};
int isEmpty(struct queue*q){
    return q->front>q->rear;
}
void enqueue(struct queue*q,int data){
    q->arr[++q->rear]=data;
}
int deque(struct queue*q){
    if(isEmpty(q))return -1;
    return q->arr[q->front++];
}
struct graph *creategraph(int numsize){
    struct graph *graph=(struct graph*)malloc(sizeof(struct graph));
    graph->numsize=numsize;
    graph->adjlist=(struct node**)malloc(sizeof(struct node*)*numsize);
    graph->visited=(int*)malloc(sizeof(int)*numsize);
    for(int i=0;i<numsize;i++){
        graph->adjlist[i]=NULL;
        graph->visited[i]=0;
    }
    return graph;
}
void addEdge(struct graph *graph,int src,int dst){
    struct node *newnode=createnode(src);
    newnode->next=graph->adjlist[dst];
    graph->adjlist[dst]=newnode;
    newnode=createnode(dst);
    newnode->next=graph->adjlist[src];
    graph->adjlist[src]=newnode;
}
void bfs(struct graph *graph,int src){
    struct queue*q=createque(graph->numsize);
    graph->visited[src]=1;
    enqueue(q,src);
    while(!isEmpty(q)){
        int n=deque(q);
        printf("%d ",n);
        struct node*temp=graph->adjlist[n];
        while(temp!=NULL){
            int t=temp->data;
            if(graph->visited[t]==0){
                graph->visited[t]=1;
                enqueue(q,t);
            }
            temp=temp->next;
        }
    }

}
int main(){
    struct graph *graph=creategraph(5);
    addEdge(graph,0,1);
    addEdge(graph,1,2);
    addEdge(graph,2,3);
    addEdge(graph,3,4);
    addEdge(graph,0,3);
    bfs(graph,0);
    return 0;
}

File: python/S15_GRAPH/L1_BASIC/4_bfs.py
class Node:
    def __init__(self,data):
        self.data=data
        self.next=None
class queue:
    def __init__(self, num):
        self.rear = 0
        self.front = 0
        self.arr = [0 for _ in range(num)]

    def isEmpty(self):
        return self.front == self.rear

    def enqueue(self, data):
        self.arr[self.rear] = data
        self.rear += 1

    def dequeue(self):
        if self.isEmpty():
            return -1
        p = self.arr[self.front]
        self.front += 1
        return p


class graph:
    def __init__(self, num):
        self.num = num
        self.adjlist = [None] * num
        self.visited = [0 for _ in range(num)]

    def addEdge(self, src, dst):
        newnode = Node(dst)
        newnode.next = self.adjlist[src]
        self.adjlist[src] = newnode

        newnode = Node(src)
        newnode.next = self.adjlist[dst]
        self.adjlist[dst] = newnode

    def bfs(self, src):
        q = queue(self.num)

        self.visited[src] = 1
        q.enqueue(src)

        while not q.isEmpty():
            t = q.dequeue()
            print(t, end=" ")

            temp = self.adjlist[t]
            while temp is not None:
                n = temp.data
                if not self.visited[n]:
                    self.visited[n] = 1
                    q.enqueue(n)
                temp = temp.next

if __name__=="__main__":
    g=graph(5)
    g.addEdge(0,1)
    g.addEdge(1,2)
    g.addEdge(2,3)
    g.addEdge(3,4)
    g.bfs(0)
